File: packages/plugin-swc/src/ast_transformers.rs
/**
 * KEEP IN SYNC:
 * - `packages/plugin-tsc/src/ast-transformers.ts`
 * - `packages/plugin-swc/src/ast_transformers.rs`
 */
use swc_core::ecma::ast::{self as t};

pub fn transform_import_declaration(node: &t::ImportDecl) -> Option<t::ImportDecl> {
  if node.src.value.contains("sayable") {
    return Some(transform_macro_import_declaration(node));
  }

  None
}

fn transform_macro_import_declaration(node: &t::ImportDecl) -> t::ImportDecl {
  let specifier = match &*node.src.value {
    "sayable" => "sayable/runtime",
    "@sayable/react" => "@sayable/react/runtime",
    _ => node.src.value.as_str(),
  };

  t::ImportDecl {
    src: Box::new(t::Str {
      span: node.src.span,
      value: specifier.into(),
      raw: None,
    }),
    ..node.clone()
  }
}

File: packages/plugin-tsc/src/ast-transformers.ts
/**
 * KEEP IN SYNC:
 * - `packages/plugin-tsc/src/ast-transformers.ts`
 * - `packages/plugin-swc/src/ast_transformers.rs`
 */

import t, { factory as f } from 'typescript';

/**
 * Transforms imports from `sayable` or `@sayable/react` into their runtime equivalents.
 *
 * Only matches imports where the module specifier includes `"sayable"`.
 *
 * @param node The original import declaration node
 * @returns A transformed import declaration, or null if not matched
 */
export function transformImportDeclaration(node: t.ImportDeclaration) {
  if (
    t.isStringLiteral(node.moduleSpecifier) &&
    node.moduleSpecifier.text.includes('sayable')
  ) {
    return transformMacroImportDeclaration(node);
  }

  return null;
}

/**
 * Rewrites `sayable` and `@sayable/react` imports to their `/runtime` equivalents.
 *
 * @param node The import declaration node to transform
 * @returns A new import declaration node with updated module specifier
 */
function transformMacroImportDeclaration(node: t.ImportDeclaration) {
  let specifier = (node.moduleSpecifier as t.StringLiteral).text;
  if (specifier === 'sayable') specifier = 'sayable/runtime';
  if (specifier === '@sayable/react') specifier = '@sayable/react/runtime';

  return f.updateImportDeclaration(
    node,
    node.modifiers,
    node.importClause,
    f.createStringLiteral(specifier),
    node.attributes,
  );
}
